### ISSUE TITLE
Model Caching with Incremental Learning - 5-50x Speedup for Repeated Analyses

### ISSUE THREAD

**Issue by alexoms** [OptimalMatch/temporal-trading-agents#62]
## Overview
Implemented comprehensive model caching system with incremental learning (fine-tuning) to dramatically reduce analysis time for repeated analyses. Instead of training from scratch every time, the system now:
1. Caches trained models
2. Reuses cached models when fresh enough
3. Fine-tunes with new data when needed
4. Automatically retrains when models become stale

### Before (No Caching)
- **Every analysis:** 15-25 minutes
- **Repeated analysis (same symbol):** 15-25 minutes (wasteful retraining)
- **User experience:** Long wait times for each request

### After (With Caching)

| Scenario | Time | Speedup | Notes |
|----------|------|---------|-------|
| First analysis (cold cache) | 15-25 min | 1x (baseline) | Full training, saves to cache |
| Analysis <1h old | ~30 sec | **30-50x** | Loads cached model, no training |
| Analysis 1-6h old | 2-5 min | **5-10x** | Fine-tunes 3 epochs instead of 15 |
| Analysis >6h old | 15-25 min | 1x | Retrains from scratch (data too old) |

### New Files
- **strategies/model_cache.py** (320 lines): Complete caching infrastructure

### Modified Files
- **examples/crypto_ensemble_forecast.py**: Training with cache support
- **strategies/strategy_utils.py**: Pass-through cache parameters
- **backend/main.py**: Worker function updates

### Cache Storage
```
/tmp/model_cache/
├── BTC-USD_1h_momentum_abc123_model.pt      # PyTorch model weights
├── BTC-USD_1h_momentum_abc123_scaler.pkl    # StandardScaler
└── BTC-USD_1h_momentum_abc123_meta.json     # Metadata
```

### 1. Intelligent Cache Invalidation

**Time-Based Staleness**
- Models older than 6 hours (configurable) are considered stale
- Prevents using outdated models on fast-moving markets

**Data-Driven Staleness**
- Checks if significant new data is available
- Hourly data: >2 hours of new data → retrain
- Daily data: >24 hours of new data → retrain

**Feature Mismatch Detection**
- Automatically retrains if feature columns change
- Ensures model compatibility with current data pipeline

### 2. Incremental Learning (Fine-Tuning)

When cached model is fresh but some new data exists:
- **Fine-tune for 3 epochs** instead of full 15 epochs
- **Lower learning rate** (1e-5 vs 1e-4) for stability
- **Preserves learned patterns** while adapting to new market conditions
- **Tracks cumulative epochs** across fine-tuning sessions

### 3. Configurable Parameters

```python
use_cache=True              # Enable/disable caching
max_cache_age_hours=6.0     # Max age before retrain
fine_tune_epochs=3          # Epochs for fine-tuning
```

### 4. Comprehensive Metadata Tracking

```json
{
  "symbol": "BTC-USD",
  "interval": "1h",
  "lookback": 45,
  "focus": "momentum",
  "training_timestamp": "2025-11-09T01:30:00",
  "data_end_timestamp": "2025-11-09T01:00:00",
  "training_epochs": 15,
  "best_val_loss": 0.0234,
  "model_architecture": {
    "d_model": 512,
    "num_encoder_layers": 6,
    "num_decoder_layers": 6
  }
}
```

### Visual Indicators
- 💾 Model caching enabled
- 🔄 Using cached model
- 🔧 Fine-tuning cached model
- 🏋️ Training from scratch
- 📂 Loaded model from cache (age: X.Xh)

### Example Output
```
💾 Model caching enabled (max age: 6.0h, fine-tune: 3 epochs)
📂 Loaded model from cache: BTC-USD_1h_momentum_abc123 (age: 2.3h)
🔧 Fine-tuning cached model (3 epochs)...
✓ Best validation loss: 0.0198 (fine-tuned)
💾 Saved model to cache: BTC-USD_1h_momentum_abc123
```

### Test 1: First Analysis (Cold Cache)
1. Run analysis for BTC-USD hourly
2. Expected: 15-25 minutes (full training)
3. Verify: Models saved to `/tmp/model_cache/`

### Test 2: Immediate Re-Analysis (<1h)
1. Run analysis for BTC-USD hourly again within 1 hour
2. Expected: ~30 seconds (load cached, no training)
3. Verify: Logs show "Using cached model"

### Test 3: Analysis After New Data (1-6h)
1. Wait 2-3 hours
2. Run analysis for BTC-USD hourly
3. Expected: 2-5 minutes (fine-tune 3 epochs)
4. Verify: Logs show "Fine-tuning cached model"

### Test 4: Stale Cache (>6h)
1. Wait 7+ hours
2. Run analysis for BTC-USD hourly
3. Expected: 15-25 minutes (full retrain)
4. Verify: Logs show "Training from scratch"

### Test 5: Different Symbol (Cache Miss)
1. Run analysis for ETH-USD hourly
2. Expected: 15-25 minutes (first time for this symbol)
3. Verify: New cache files created for ETH-USD

### Why /tmp/model_cache/?
- Fast access (in-memory filesystem on many systems)
- Automatic cleanup on system restart (prevents stale cache buildup)
- No database overhead
- Easy to inspect and debug

### Why 6-Hour Default Cache Age?
- Balances freshness vs performance
- Crypto markets: meaningful changes in 6 hours
- Most users analyze same symbols multiple times per day
- Can be adjusted per deployment

### Why 3 Epochs for Fine-Tuning?
- Sufficient to adapt to new patterns
- Fast enough to provide speedup
- Prevents overfitting to very recent data
- Lower learning rate reduces risk of catastrophic forgetting

### Why Lower Learning Rate for Fine-Tuning?
- Learning rate: 1e-4 → 1e-5 (10x reduction)
- Preserves existing learned representations
- Makes smaller, careful updates
- Standard practice in transfer learning

### Potential Improvements
1. **MongoDB GridFS Storage**
   - Persistent cache across container restarts
   - Shared cache across multiple backend instances
   - Cache size management and LRU eviction

2. **Cache Warming**
   - Pre-train models for popular symbols
   - Background job to keep cache fresh
   - Scheduled retraining during off-peak hours

3. **API Parameter**
   - `force_retrain=true` query parameter
   - Allows users to explicitly request fresh training
   - Useful for debugging or after major market events

4. **Cache Analytics**
   - Track cache hit/miss rates
   - Monitor average speedup
   - Alert on stale cache issues

5. **Distributed Caching**
   - Redis or Memcached for multi-instance deployments
   - Model versioning and compatibility checks
   - Automatic cache synchronization

## Related Work
- Addresses user question: "why go through the trouble of training in analysis phase and then have to redo it again?"
- Implements Option 3: Incremental Learning with caching
- Complements GPU optimization (Issue #61) for maximum performance

## Deployment Status
✅ Committed: 5d27c28  
✅ Deployed to main  
✅ Backend redeployed with caching enabled  
⏳ Ready for user testing  

## Documentation
See commit 5d27c28 for full implementation details and inline code documentation.
